lib/PrintAsObjC/PrintAsObjC.cpp

```diff
@@ -1579,62 +1580,65 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
     printNullability(optionalKind);
   }
 
-  void visitProtocolType(ProtocolType *PT, 
-                         Optional<OptionalTypeKind> optionalKind, 
-                         bool isMetatype = false) {
-    auto proto = PT->getDecl();
-    if (proto->isSpecificProtocol(KnownProtocolKind::Error)) {
+  void visitExistentialType(Type T,
+                            Optional<OptionalTypeKind> optionalKind,
+                            bool isMetatype = false) {
+    auto layout = T->getExistentialLayout();
+    if (layout.isErrorExistential()) {
       if (isMetatype) os << "Class";
       else os << "NSError *";
       printNullability(optionalKind);
       return;
     }
 
-    os << (isMetatype ? "Class" : "id");
-
-    assert(proto->isObjC());
-    if (auto knownKind = proto->getKnownProtocolKind()) {
-      if (*knownKind == KnownProtocolKind::AnyObject) {
-        printNullability(optionalKind);
-        return;
+    if (layout.superclass) {
+      auto *CD = layout.superclass->getClassOrBoundGenericClass();
+      if (isMetatype) {
+        if (CD->isObjC())
+          os << "SWIFT_METATYPE(" << getNameForObjC(CD) << ")";
+        else
+          os << "Class";
```

Comment:
Probably never, o was cargo culting the Metatype case where we seem to print non-@objc metatypes as Class.